Implement memory initialization state copy functionality

artemagvanian opened the pull request:
This PR adds support of copying memory initialization state without checks in-between. Every time a copy is performed, the tracked byte is non-deterministically switched.

Resolves #3347

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented on library/kani/src/mem_init.rs:
```diff
@@ -89,6 +89,33 @@ impl MemoryInitializationState {
+    /// adjusting the tracked offset.
+    pub fn copy<const LAYOUT_SIZE: usize>(
+        &mut self,
+        from_ptr: *const u8,
+        to_ptr: *const u8,
+        num_elts: usize,
```
Another option would be to change these interfaces to take `T` and use `size_of::<T>::()` which is cleaner.

artemagvanian replied:
Having the types to be `*const u8` is consistent with the other methods. They are not exposed in the public interface, but perhaps we should change all of them in another PR.

celinval reviewed (commented):
As discussed offline, copying can trigger a delayed UB in the presence of pointer casting from types with different padding bytes location. Can you please mitigate this case here though? We could encode copy operations as unsupported. Or we could also keep a global variable that we set whenever an incompatible cast is performed. Whenever we run `copy` or `copy_nonoverlapping`, we can assert that global variable hasn't been set.

artemagvanian commented:
I am considering waiting until we make some progress with supporting delayed UB triggered by pointer casting with different padding and revisiting this PR then

artemagvanian commented:
> Can you please add a test that has a more complex CFG / Call graph combination? Maybe the reference comes from a different function than the one that creates the raw pointer and does the copy? Some branches might be nice too. :)

That's a good idea. I added a test with a more complex call and control flow structure. I also added a couple of tests checking that this interacts well with delayed UB detection.
